examples/hitl/rearrange_v2/object_state_manipulator.py

```diff
@@ -0,0 +1,172 @@
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Tuple, cast
+
+from habitat_llm.sims.collaboration_sim import CollaborationSim
+from habitat_llm.tools.motor_skills.object_states.oracle_clean_skills import (
+    OracleCleanInPlaceSkill,
+)
+from habitat_llm.tools.motor_skills.object_states.oracle_fill_skills import (
+    OracleFillInPlaceSkill,
+)
+from habitat_llm.tools.motor_skills.object_states.oracle_object_state_skill import (
+    OracleObjectStateInPlaceSkill,
+)
+from habitat_llm.tools.motor_skills.object_states.oracle_pour_skills import (
+    OraclePourInPlaceSkill,
+)
+from habitat_llm.tools.motor_skills.object_states.oracle_power_skills import (
+    OraclePowerOffInPlaceSkill,
+    OraclePowerOnInPlaceSkill,
+)
+from world import World
+
+from habitat.sims.habitat_simulator import sim_utilities
+from habitat.sims.habitat_simulator.object_state_machine import (
+    ObjectStateSpec,
+    set_state_of_obj,
+)
+
+BooleanActionMap = Dict[bool, OracleObjectStateInPlaceSkill]
+
+
+@dataclass
+class BooleanAction:
+    state_spec: ObjectStateSpec
+    current_value: bool
+    target_value: bool
+    available: bool
+    enabled: bool
+    error: Optional[str]
+
+
+class ObjectStateManipulator:
+    """
+    Helper class for reading and writing object states for a given world, on behalf of a given agent.
+    Requires the 'habitat_llm' external module.
+    """
+
+    def __init__(
+        self,
+        sim: CollaborationSim,
+        agent_index: int,
+        world: World,
+        maximum_distance: float,
+    ):
+        self._sim = sim
+        self._agent_index = agent_index
+        self._maximum_distance = maximum_distance
+        self._world = world
+
+        self._boolean_state_action_map: Dict[str, BooleanActionMap] = {
+            "is_clean": {True: OracleCleanInPlaceSkill},
+            "is_filled": {
+                True: OracleFillInPlaceSkill,
+                False: OraclePourInPlaceSkill,
+            },
+            "is_powered_on": {
+                True: OraclePowerOnInPlaceSkill,
+                False: OraclePowerOffInPlaceSkill,
+            },
+        }
```

Comment:
I think you would reverse engineer the ObjectStateMachine instance by checking each active ObjectStateSpec for it's name: https://github.com/facebookresearch/habitat-lab/blob/cf2a6aa76f8081853b1b4be3844f12fdbfd95d83/habitat-lab/habitat/sims/habitat_simulator/object_state_machine.py#L230

Could be a helper function to do this in that class if it would be useful.

Reply:
Sounds like what we want. That may require some `habitat_llm` changes for skill auto-discovery.
I'll refactor this in a later pass.